cherab/solps/models/line_emitter.pyx

```diff
@@ -0,0 +1,114 @@
+# cython: language_level=3
+
+# Copyright 2016-2021 Euratom
+# Copyright 2016-2021 United Kingdom Atomic Energy Authority
+# Copyright 2016-2021 Centro de Investigaciones Energéticas, Medioambientales y Tecnológicas
+#
+# Licensed under the EUPL, Version 1.1 or – as soon they will be approved by the
+# European Commission - subsequent versions of the EUPL (the "Licence");
+# You may not use this work except in compliance with the Licence.
+# You may obtain a copy of the Licence at:
+#
+# https://joinup.ec.europa.eu/software/page/eupl5
+#
+# Unless required by applicable law or agreed to in writing, software distributed
+# under the Licence is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR
+# CONDITIONS OF ANY KIND, either express or implied.
+#
+# See the Licence for the specific language governing permissions and limitations
+# under the Licence.
+
+from raysect.optical cimport Spectrum, Point3D, Vector3D
+from cherab.core cimport Plasma, AtomicData
+from cherab.core.model.lineshape cimport GaussianLine, LineShapeModel
+from cherab.core.utility.constants cimport RECIP_4_PI
+from cherab.core.math.mappers import AxisymmetricMapper
+
+from cherab.solps cimport SOLPSFunction2D
+
+
+cdef class SOLPSLineEmitter(PlasmaModel):
+    """
+    Line emitter that uses a custom `SOLPSFunction2D` function for radiation power density.
+
+    :param SOLPSFunction2D radiation_function: The power density function defined on the SOLPS mesh.
+    :param Line line: Spectral line.
+    :param Plasma plasma: Plasma object. Default is None.
+    :param AtomicData atomic_data: Atomic data provider. Will use `OpenADAS` if None (default).
```

Comment:
I can't find where you specify the OpenADAS() when None is passed, is this accurate?

Reply:
This is a good question. When the model is added to the `plasma.models`, the `atomic_data` of that plasma object will be used. This is invoked in `Plasma._configure_geometry()` and is done by `PlasmaMaterial` initialiser:
https://github.com/cherab/core/blob/0eb7454137b17bfbeca6d5e3eae4281f9cd4c157/cherab/core/plasma/material.pyx#L25-L46. 
So, it does not matter what values for `plasma` and `atomic_data` are provided at initialisation, they will be overwritten, as soon as the model is added to `Plasma.models`. The only way to use the `atomic_data` different from that of the plasma object, to which the model is added via the `models` list, is by setting `atomic_data` once the model is already connected to the plasma. However, one can use the `PlasmaModel` without adding it to the `models` list, for example to sample the emission spectrum, and this can be done with the `atomic_data` provided at initialisation.

The `atomic_data` is not set to `OpenADAS ()` by default in either `Plasma` or any implemented `PlasmaModel`. I am just following this pattern here. I suspect this goes back to the days when `cherab.openadas` was distributed separately from Cherab, and now this can be fixed. But, if we are going to fix this, I think it should be done in the parent class.

Reply:
Right, sorry I missed this. Thanks for pointing it out @vsnever